hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HStore.java

```diff
@@ -246,6 +234,8 @@
   private AtomicLong compactedCellsSize = new AtomicLong();
   private AtomicLong majorCompactedCellsSize = new AtomicLong();
 
+  private HStoreContext storeContext;
```

Comment:
Why introduce an StoreContext? Isn't that what a Store is? Store spans files.

(We need the 'H' in HStoreContext?)

Reply:
Yeah, just trying to understand why. You want to swap out the HStore implementation or something?

Reply:
we introduced this context for a reason. `HStore` is a big informative object passed as a input to many file operation classes e.g. `StoreFlusher` to get access for related information in reference, e.g. `columnFamilyDescriptor` and `regionFileSystem`, we're trying to group those `read-only` reference such that we don't have to always use `HStore` object in the lower classes. 

Although this refactoring is base for the upcoming patches in HBASE-24749 that we will introduce new `StoreFileWriterFactory` and `StoreFileCommitter`, we think this could be a good wrapper to group and limit access for internal reference as well. 

> (We need the 'H' in HStoreContext?)

we can remove the `H` from naming, will do in the next commit

Reply:
Yeah the idea is that HStore has way more scope than most consumers of HStore actually need (1. The information about the store and 2. the functions to mutate things - A lot of callers just care about the info, not the actual functions) so it made sense to separate out the info to a POJO to reduce the scope of what is being exposed.

Reply:
Ok. Stick that on the head of the context class justifying why it exists.

Where do you draw the line on what is in context and what is in Store? Thanks.

Reply:
IMO those informative accessor/reference and final/read-only primitives should ideally be in the context, although we're focusing on writer (`StoreFileWriter`) related reference and may have missed few of them (e.g. `scanInfo`) in this commit. 

let's try to clarify your suggestion 
1. if you see the `StoreContext` is general to be applied on most cases, are those missing fields (e.g. `scanInfo` and  final primitives) what you're trying to point out ? if so, we can revisit and filter/add more into the `StoreContext` 
2. The scope of this `Context` is more related to Writer(`StoreFileWriter`)/Committer(will be added), should we rename it to `StoreWriterContext`/`StoreWriteContext`  that used by those operators?

Reply:
@saintstack , do you have other comments about the scope of this StoreContext ? do we need to rename it ? or do you think we could move forward and introduce this context object with the proposed set of informative accessor/reference ?

Reply:
@taklwu

Sorry for late reply (Thanks for pinging).

> let's try to clarify your suggestion

I was asking a question, not making a suggestion -- smile.

One thing I notice is that you and @z-york talk of the Store 'info' or 'information'. So, would StoreInfo make more sense than StoreContext? It would align with ScanInfo (Yeah, StoreInfo looks like it should have ScanInfo since it a superset or should subsume ScanInfo).

Above you say this... "...we're trying to group those read-only reference such that we don't have to always use HStore object in the lower classes."  Seems like you could use a version of this to answer my question (ScanInfo is 'Immutable information for scans over a store'... so StoreInfo could be Store immutable info?).

Back to your comments...

I was trying to figure when to pass Store and when I'd pass StoreInfo/StoreContext only.

Lets just have a StoreContext/StoreInfo. Lets NOT have a StoreWriterContext; i.e. a context for write-side only.

One sec.... let me look at the PR here.

Reply:
thanks, I should have provide a followup in the latest diff and proposed to keep with the `StoreContext` naming.

marked this conversation as resolved.